Deadlock fixes

cstamas opened the pull request:
Step1: Change to DefaultInspectorHost wrt async inspector handling.

Before this, the event inspectors were processed in undefined order,
as they were coming in (as SISU injected them, unsure was this
nondeterministic or not). Anyway, event inspector implementations
was impled with NO expectations about order, so this change is fine.

With this change, all that happens is that system GUARANTEES
that when an async inspector is about to be executed, all the
sync inspectors were already run and finished.

Hence, this change makes fix for NEXUS-4784 simple: just factor out the
task scheduling code after line (currently 130)

getIndexerManager().updateRepositoryIndexContext( repository.getId() );

(but keep this line above in sync inspector), and this fixes
the problem described in NEXUS-4784 (and many others too).

Step2: Factor out task scheduling code from any sync EventInspector into async EventInspectors.

Step3: Revisit DefaultScheduler (component doing task scheduling) and evaluate use of ConcurrentHashMap and other better solutions over current strict locking.
# 

In progress, not for vote yet.

cstamas committed:
Change to DefaultInspectorHost wrt async inspector handling.

Before this, the event inspectors were processed in undefined order,
as they were coming in (as SISU injected them, unsure was this
nondeterministic or not). Anyway, event inspector implementations
was impled with NO expectations about order, so this change is fine.

With this change, all that happens is that system GUARANTEES
that when an async inspector is about to be executed, all the
sync inspectors were already run and finished.

Hence, this change makes fix for NEXUS-4784 simple: just factor out the
task scheduling code after line (currently 130)

getIndexerManager().updateRepositoryIndexContext( repository.getId() );

(but keep this line above in sync inspector), and this fixes
the problem described in NEXUS-4784 (and many others too).

cstamas commented:
The sync event inspector mentioned in this pull (and in issue NEXUS-4784) is:
https://github.com/sonatype/nexus/blob/master/nexus/nexus-core-plugins/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/events/IndexingRepositoryRegistryRepositoryEventInspector.java#L128

mcculls commented on nexus/nexus-app/src/main/java/org/sonatype/nexus/events/DefaultEventInspectorHost.java:
```diff
@@ -111,30 +111,53 @@ protected void processEvent( final Event<?> evt )
+                {
+                    getLogger().warn(
+                        "Async EventInspector implementation='" + ei.getClass().getName()
+                            + "' had problem accepting an event='" + evt.getClass() + "'", e );
```
Should this use the same message format style as the other logging calls in this class?

cstamas replied:
Fixed in subsequent commit.

mcculls commented:
+1 on implementation, but should also add a test to verify that all synchronous handlers run before asynchronous ones to avoid future regressions.

mcculls commented:
&lt;off-topic&gt;
Wrt. component ordering - it is deterministic in that repeated executions on the same machine will be consistent, but the underlying sequence is effectively based on the generated app-booter classpath (which in turn is based on the file-system ordering) and the order that plugins are loaded. That's why plexus (and sisu running in plexus-mode) made no guarantees about ordering in general.

However with sisu you can enforce an ordering using the RankingFunction SPI - for example the default plexus ordering makes sure "default" hinted components appear before all others - but any ordering function is possible using this SPI.
&lt;/off-topic&gt;

For this issue, the two-pass approach is the best solution because it's simple and effective.

cstamas committed:
Applied Stuart's remarks and added UT to prevent future regressions.

cstamas commented:
Added UT to protect from future regressions.

cstamas committed:
Making test EI "more correct", latch is released as very last step.

mah0526 commented:
+1

cstamas committed:
Applied comments from Mike, also simplifying component.

cstamas committed:
NEXUS-4744: factoring out the background task spawning code

cstamas committed:
SISU Goodies not on compile classpath

Reverting to Guice anno for now, later we should consider putting
sisu goodies onto compile classpath (is on test currently and eclipse
sadly does not support "scopes")

nabcos commented on nexus/nexus-core-plugins/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/events/IndexingRepositoryRegistryRepositoryAsyncEventInspector.java:
```diff
@@ -0,0 +1,190 @@
+                // a) if remoteUrl changed
+                // b) if download remote index enabled (any repo type)
+                // c) if repository is made searchable
+                // TODO: are we sure only a) needs a check for Maven2? I think all of them need
```
TODO seems out of place, repository is not checked at all AFAICS

cstamas commented:
CI builds passed:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/182/

Note: failures were due to missing license headers of newly added files, check runs AFTER UTs/ITs

bdemers commented:
+1

cstamas committed:
Added license header to file this branch introduces.

For other two, probably the master branch when it was branched off contained these files.

nabcos commented:
+1

cstamas commented:
There are 4 :+1:'s, plus the CI :+1: so am merging this change in to make progress faster wrt Scheduler (we can still leave out completely Scheduler changes if we see them dangerous to make them now, Scheduler is still unchanged so far, with this pull req inclusive)
